File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/novel/pages/NovelFragmentChapters.java
package com.github.doomsdayrs.apps.shosetsu.ui.novel.pages;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.github.Doomsdayrs.api.shosetsu.services.core.objects.NovelChapter;
import com.github.doomsdayrs.apps.shosetsu.R;
import com.github.doomsdayrs.apps.shosetsu.backend.Download_Manager;
import com.github.doomsdayrs.apps.shosetsu.backend.Utilities;
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database;
import com.github.doomsdayrs.apps.shosetsu.ui.novel.NovelFragment;
import com.github.doomsdayrs.apps.shosetsu.ui.novel.adapters.ChaptersAdapter;
import com.github.doomsdayrs.apps.shosetsu.ui.novel.async.ChapterLoader;
import com.github.doomsdayrs.apps.shosetsu.variables.DownloadItem;
import com.github.doomsdayrs.apps.shosetsu.variables.enums.Status;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;

import static com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseIdentification.getChapterIDFromChapterURL;

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 * Shosetsu
 * 9 / June / 2019
 *
 * @author github.com/doomsdayrs
 */

/**
 * Displays the chapters the novel contains
 * TODO Check filesystem if the chapter is saved, even if not in DB.
 */
public class NovelFragmentChapters extends Fragment {

    public ArrayList<NovelChapter> selectedChapters = new ArrayList<>();

    public boolean contains(NovelChapter novelChapter) {
        for (NovelChapter n : selectedChapters)
            if (n.link.equalsIgnoreCase(novelChapter.link))
                return true;
        return false;
    }

    private int currentMaxPage = 1;

    private int findMinPosition() {
        int min = novelFragment.novelChapters.size();
        for (int x = 0; x < novelFragment.novelChapters.size(); x++)
            if (contains(novelFragment.novelChapters.get(x)))
                if (x < min)
                    min = x;
        return min;
    }

    public static boolean reversed;
    @SuppressLint("StaticFieldLeak")
    public static RecyclerView recyclerView;

    private int findMaxPosition() {
        int max = -1;
        for (int x = novelFragment.novelChapters.size() - 1; x >= 0; x--)
            if (contains(novelFragment.novelChapters.get(x)))
                if (x > max)
                    max = x;
        return max;
    }

    public static ChaptersAdapter adapter;
    public SwipeRefreshLayout swipeRefreshLayout;
    public NovelFragment novelFragment;
    public TextView pageCount;
    public FloatingActionButton resumeRead;

    /**
     * Constructor
     */
    public NovelFragmentChapters() {
        setHasOptionsMenu(true);
    }

    public void setNovelFragment(NovelFragment novelFragment) {
        this.novelFragment = novelFragment;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        reversed = false;
        Log.d("NFChapters", "Destroy");
        recyclerView = null;
        adapter = null;
    }

    /**
     * Save data of view before destroyed
     *
     * @param outState output save
     */
    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putInt("maxPage", currentMaxPage);
        outState.putSerializable("selChapter", selectedChapters);
    }

    /**
     * Creates view
     *
     * @param inflater           inflater to retrieve objects
     * @param container          container of this fragment
     * @param savedInstanceState save
     * @return View
     */
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            selectedChapters = (ArrayList<NovelChapter>) savedInstanceState.getSerializable("selChapter");
            currentMaxPage = savedInstanceState.getInt("maxPage");
        }
        Log.d("NovelFragmentChapters", "Creating");
        View view = inflater.inflate(R.layout.fragment_novel_chapters, container, false);
        recyclerView = view.findViewById(R.id.fragment_novel_chapters_recycler);
        swipeRefreshLayout = view.findViewById(R.id.fragment_novel_chapters_refresh);
        pageCount = view.findViewById(R.id.page_count);
        resumeRead = view.findViewById(R.id.resume);
        resumeRead.setVisibility(View.GONE);
        swipeRefreshLayout.setOnRefreshListener(() -> new ChapterLoader(novelFragment.novelPage, novelFragment.novelURL, novelFragment.formatter).setNovelFragmentChapters(this).execute(getActivity()));

        if (savedInstanceState != null) {
            currentMaxPage = savedInstanceState.getInt("maxPage");
        }
        setChapters();
        onResume();
        resumeRead.setOnClickListener(view1 -> {
            int i = novelFragment.lastRead();
            if (i != -1 && i != -2)
                Utilities.openChapter(getActivity(), novelFragment.novelChapters.get(i), novelFragment.novelID, novelFragment.formatter.getID());
            else
                Toast.makeText(getContext(), "No chapters! How did you even press this!", Toast.LENGTH_SHORT).show();
        });
        return view;
    }

    /**
     * Sets the novel chapters down
     */
    public void setChapters() {
        recyclerView.post(() -> {
            recyclerView.setHasFixedSize(false);
            RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext());
            if (Database.DatabaseNovels.inDatabase(novelFragment.novelID)) {
                novelFragment.novelChapters = Database.DatabaseChapter.getChapters(novelFragment.novelID);
                if (novelFragment.novelChapters != null && novelFragment.novelChapters.size() != 0)
                    resumeRead.setVisibility(View.VISIBLE);
            }
            adapter = new ChaptersAdapter(this);
            adapter.setHasStableIds(true);
            recyclerView.setLayoutManager(layoutManager);
            recyclerView.setAdapter(adapter);
        });
    }

    public Menu menu;

    public MenuInflater getInflater() {
        return new MenuInflater(getContext());
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        switch (item.getItemId()) {
            case R.id.chapter_select_all:
                for (NovelChapter novelChapter : novelFragment.novelChapters)
                    if (!contains(novelChapter))
                        selectedChapters.add(novelChapter);
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_download_selected:
                for (NovelChapter novelChapter : selectedChapters) {
                    int chapterID = getChapterIDFromChapterURL(novelChapter.link);
                    if (!Database.DatabaseChapter.isSaved(chapterID)) {
                        DownloadItem downloadItem = new DownloadItem(novelFragment.formatter, novelFragment.novelPage.title, novelChapter.title, chapterID);
                        Download_Manager.addToDownload(downloadItem);
                    }
                }
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_delete_selected:
                for (NovelChapter novelChapter : selectedChapters) {
                    int chapterID = getChapterIDFromChapterURL(novelChapter.link);
                    if (Database.DatabaseChapter.isSaved(chapterID))
                        Download_Manager.delete(getContext(), new DownloadItem(novelFragment.formatter, novelFragment.novelPage.title, novelChapter.title, chapterID));
                }
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_deselect_all:
                selectedChapters = new ArrayList<>();
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                onCreateOptionsMenu(menu, getInflater());
                return true;

            case R.id.chapter_mark_read:
                for (NovelChapter novelChapter : selectedChapters) {
                    int chapterID = getChapterIDFromChapterURL(novelChapter.link);

                    if (Database.DatabaseChapter.getStatus(chapterID).getA() != 2)
                        Database.DatabaseChapter.setChapterStatus(chapterID, Status.READ);
                }
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_mark_unread:
                for (NovelChapter novelChapter : selectedChapters) {
                    int chapterID = getChapterIDFromChapterURL(novelChapter.link);

                    if (Database.DatabaseChapter.getStatus(chapterID).getA() != 0)
                        Database.DatabaseChapter.setChapterStatus(chapterID, Status.UNREAD);
                }
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_mark_reading:
                for (NovelChapter novelChapter : selectedChapters) {
                    int chapterID = getChapterIDFromChapterURL(novelChapter.link);
                    if (Database.DatabaseChapter.getStatus(chapterID).getA() != 0)
                        Database.DatabaseChapter.setChapterStatus(chapterID, Status.READING);
                }
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_select_between:
                int min = findMinPosition();
                int max = findMaxPosition();
                for (int x = min; x < max; x++)
                    if (!contains(novelFragment.novelChapters.get(x)))
                        selectedChapters.add(novelFragment.novelChapters.get(x));
                NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
                return true;

            case R.id.chapter_filter:
                Collections.reverse(novelFragment.novelChapters);
                NovelFragmentChapters.reversed = !NovelFragmentChapters.reversed;
                return NovelFragmentChapters.recyclerView.post(() -> NovelFragmentChapters.adapter.notifyDataSetChanged());
        }
        return false;
    }

    /**
     * Creates the option menu (on the top toolbar)
     *
     * @param menu     Menu reference to fill
     * @param inflater Object to inflate the menu
     */
    @Override
    public void onCreateOptionsMenu(@NotNull Menu menu, @NotNull MenuInflater inflater) {
        this.menu = menu;
        menu.clear();
        if (selectedChapters.size() <= 0)
            inflater.inflate(R.menu.toolbar_chapters, menu);
        else
            inflater.inflate(R.menu.toolbar_chapters_selected, menu);
    }
}

File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/novel/pages/NovelFragmentChapters.kt
package com.github.doomsdayrs.apps.shosetsu.ui.novel.pages

import android.app.AlertDialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.*
import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.github.doomsdayrs.api.shosetsu.services.core.Novel
import com.github.doomsdayrs.apps.shosetsu.R
import com.github.doomsdayrs.apps.shosetsu.backend.DownloadManager
import com.github.doomsdayrs.apps.shosetsu.backend.Utilities
import com.github.doomsdayrs.apps.shosetsu.backend.Utilities.openChapter
import com.github.doomsdayrs.apps.shosetsu.backend.ViewedController
import com.github.doomsdayrs.apps.shosetsu.backend.async.ChapterLoader
import com.github.doomsdayrs.apps.shosetsu.backend.async.ChapterLoader.ChapterLoaderAction
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.addToChapters
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.getChapter
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.getChapterStatus
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.getChapters
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.isNotInChapters
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.isSaved
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.setChapterStatus
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseChapter.updateChapter
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseIdentification.getChapterIDFromChapterURL
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database.DatabaseUpdates.addToUpdates
import com.github.doomsdayrs.apps.shosetsu.ui.novel.NovelController
import com.github.doomsdayrs.apps.shosetsu.ui.novel.adapters.ChaptersAdapter
import com.github.doomsdayrs.apps.shosetsu.variables.DownloadItem
import com.github.doomsdayrs.apps.shosetsu.variables.enums.Status
import com.github.doomsdayrs.apps.shosetsu.variables.ext.context
import com.github.doomsdayrs.apps.shosetsu.variables.ext.getString
import com.github.doomsdayrs.apps.shosetsu.variables.ext.toast
import com.github.doomsdayrs.apps.shosetsu.variables.obj.Broadcasts
import com.google.android.material.chip.Chip
import com.google.android.material.floatingactionbutton.FloatingActionButton

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 */
/**
 * Shosetsu
 * 9 / June / 2019
 *
 * @author github.com/doomsdayrs
 *
 *
 * Displays the chapters the novel contains
 * TODO Check filesystem if the chapter is saved, even if not in DB.
 *
 */
class NovelFragmentChapters : ViewedController() {
    override val layoutRes: Int = R.layout.fragment_novel_chapters


    @Attach(R.id.resume)
    var resume: FloatingActionButton? = null
    @Attach(R.id.page_count)
    var pageCount: Chip? = null
    @Attach(R.id.fragment_novel_chapters_refresh)
    var fragmentNovelChaptersRefresh: SwipeRefreshLayout? = null
    @Attach(R.id.fragment_novel_chapters_recycler)
    var fragmentNovelChaptersRecycler: RecyclerView? = null

    private val chaptersLoadedAction: ChapterLoaderAction
    private var currentMaxPage = 1
    var selectedChapters = ArrayList<Int>()
    var adapter: ChaptersAdapter? = ChaptersAdapter(this)
    var novelFragment: NovelController? = null
    var reversed = false

    var menu: Menu? = null

    init {
        setHasOptionsMenu(true)
        chaptersLoadedAction = object : ChapterLoaderAction {
            override fun onPreExecute() {
                fragmentNovelChaptersRefresh?.isRefreshing = true
            }

            override fun onPostExecute(result: Boolean, finalChapters: ArrayList<Novel.Chapter>) {
                fragmentNovelChaptersRefresh?.isRefreshing = false

                if (result) {
                    novelFragment?.novelChapters = finalChapters
                    adapter?.notifyDataSetChanged()
                }
            }

            override fun onJustBeforePost(finalChapters: ArrayList<Novel.Chapter>) {
                val s = getString(R.string.processing_data)
                pageCount?.post { pageCount?.text = s }
                for ((count: Int, novelChapter: Novel.Chapter) in finalChapters.withIndex()) {
                    val sc = s + ": $count/${finalChapters.size}"
                    pageCount?.post { pageCount?.text = sc }
                    if (novelFragment != null && novelFragment!!.novelID != -1) {
                        if (isNotInChapters(novelChapter.link)) {
                            Log.i("ChapterLoader", "Adding ${novelChapter.link}")
                            addToChapters(novelFragment!!.novelID, novelChapter)
                            if (!novelFragment!!.new)
                                addToUpdates(novelFragment!!.novelID, novelChapter.link, System.currentTimeMillis())
                        } else {
                            updateChapter(novelChapter)
                        }
                    } else Log.e("ChapterLoader", "Invalid novelID")
                }
            }

            override fun onIncrementingProgress(page: Int, max: Int) {
                val s = "Page: $page/$max"
                pageCount?.post { pageCount?.text = s }
            }

            override fun errorReceived(errorString: String) {
                Log.e("ChapterLoader", errorString)
                activity?.runOnUiThread {
                    context?.toast(errorString)
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        reversed = false
        Log.d("NFChapters", "Destroy")
        activity?.unregisterReceiver(receiver)
    }

    override fun onViewCreated(view: View) {
        resume?.visibility = GONE
        if (novelFragment != null)
            fragmentNovelChaptersRefresh?.setOnRefreshListener {
                if (novelFragment != null && novelFragment!!.novelURL.isNotEmpty())
                    ChapterLoader(chaptersLoadedAction, novelFragment!!.formatter, novelFragment!!.novelURL).execute()
            }

        setChapters()
        resume?.setOnClickListener {
            val i = novelFragment!!.lastRead()
            if (i != -1 && i != -2) {
                if (activity != null) openChapter(activity!!, novelFragment!!.novelChapters[i], novelFragment!!.novelID, novelFragment!!.formatter.formatterID)
            } else context?.toast("No chapters! How did you even press this!")
        }
        val filter = IntentFilter()
        filter.addAction(Broadcasts.BROADCAST_NOTIFY_DATA_CHANGE)
        receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                fragmentNovelChaptersRecycler?.adapter?.notifyDataSetChanged()
            }
        }
        activity?.registerReceiver(receiver, filter)
    }

    /**
     * Save data of view before destroyed
     *
     * @param outState output save
     */
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("maxPage", currentMaxPage)
        outState.putIntegerArrayList("selChapter", selectedChapters)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        selectedChapters = savedInstanceState.getIntegerArrayList("selChapter")!!
        currentMaxPage = savedInstanceState.getInt("maxPage")
    }

    /**
     * Creates view
     *
     * @param inflater           inflater to retrieve objects
     * @param container          container of this fragment
     * @param savedInstanceState save
     * @return View
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        novelFragment = parentController as NovelController?
        novelFragment!!.novelFragmentChapters = this
        Log.d("NovelFragmentChapters", "Creating")
        return super.onCreateView(inflater, container)
    }

    /**
     * Creates the option menu (on the top toolbar)
     *
     * @param menu     Menu reference to fill
     * @param inflater Object to inflate the menu
     */
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        this.menu = menu
        menu.clear()
        if (selectedChapters.size <= 0) inflater.inflate(R.menu.toolbar_chapters, menu) else inflater.inflate(R.menu.toolbar_chapters_selected, menu)
    }

    /**
     * Sets the novel chapters down
     */
    fun setChapters() {
        fragmentNovelChaptersRecycler!!.post {
            fragmentNovelChaptersRecycler!!.setHasFixedSize(false)
            val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(context)
            if (novelFragment != null && !Database.DatabaseNovels.isNotInNovels(novelFragment!!.novelID)) {
                novelFragment!!.novelChapters = getChapters(novelFragment!!.novelID)
                if (novelFragment!!.novelChapters.isNotEmpty()) resume!!.visibility = VISIBLE
            }
            adapter = ChaptersAdapter(this)
            adapter!!.setHasStableIds(true)
            fragmentNovelChaptersRecycler!!.layoutManager = layoutManager
            fragmentNovelChaptersRecycler!!.adapter = adapter
        }
    }

    val inflater: MenuInflater?
        get() = MenuInflater(context)

    fun updateAdapter(): Boolean {
        return fragmentNovelChaptersRecycler!!.post { if (adapter != null) adapter!!.notifyDataSetChanged() }
    }

    private fun customAdd(count: Int) {

        val ten = novelFragment?.getCustom(count, object : NovelController.Custom {
            override fun customCheck(status: Status): Boolean {
                return true
            }
        })
        if (!ten.isNullOrEmpty())
            for (chapter in ten)
                DownloadManager.addToDownload(activity!!, DownloadItem(novelFragment!!.formatter, novelFragment!!.novelPage.title, chapter.title, getChapterIDFromChapterURL(chapter.link)))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.download -> {
                val builder = AlertDialog.Builder(activity!!)
                builder.setTitle(R.string.download)
                        .setItems(R.array.chapters_download_options) { _, which ->
                            when (which) {
                                0 -> {
                                    // All
                                    for (chapter in novelFragment?.novelChapters!!)
                                        DownloadManager.addToDownload(activity!!, DownloadItem(novelFragment!!.formatter, novelFragment!!.novelPage.title, chapter.title, getChapterIDFromChapterURL(chapter.link)))
                                }
                                1 -> {
                                    // Unread
                                    for (chapter in novelFragment?.novelChapters!!) {
                                        val id = getChapterIDFromChapterURL(chapter.link)
                                        if (getChapterStatus(id) == (Status.UNREAD))
                                            DownloadManager.addToDownload(activity!!, DownloadItem(novelFragment!!.formatter, novelFragment!!.novelPage.title, chapter.title, id))
                                    }
                                }
                                2 -> {
                                    // TODO Custom
                                    Utilities.regret(context!!)
                                }
                                3 -> {
                                    Log.d("NovelFragmentChapters", "Downloading next 10")
                                    customAdd(10)
                                }
                                4 -> {
                                    Log.d("NovelFragmentChapters", "Downloading next 5")
                                    customAdd(5)
                                }
                                5 -> {
                                    // Download next
                                    val last = novelFragment!!.getLastRead()
                                    val next = novelFragment!!.getNextChapter(last.link, novelFragment!!.novelChapters)
                                    if (next != null)
                                        DownloadManager.addToDownload(activity!!, DownloadItem(novelFragment!!.formatter, novelFragment!!.novelPage.title, next.title, getChapterIDFromChapterURL(next.link)))
                                }
                            }
                        }
                builder.create().show()
                true
            }
            R.id.chapter_select_all -> {
                for (novelChapter in novelFragment!!.novelChapters) if (!contains(novelChapter)) selectedChapters.add(getChapterIDFromChapterURL(novelChapter.link))
                updateAdapter()
                true
            }
            R.id.chapter_download_selected -> {
                for (chapterID in selectedChapters) {
                    val chapter = getChapter(chapterID)
                    if (!isSaved(chapterID)) {
                        val downloadItem = DownloadItem(novelFragment!!.formatter, novelFragment!!.novelPage.title, chapter!!.title, chapterID)
                        DownloadManager.addToDownload(activity, downloadItem)
                    }
                }
                updateAdapter()
                true
            }
            R.id.chapter_delete_selected -> {
                for (chapterID in selectedChapters) {
                    val chapter = getChapter(chapterID)
                    if (isSaved(chapterID)) DownloadManager.delete(context, DownloadItem(novelFragment!!.formatter, novelFragment!!.novelPage.title, chapter!!.title, chapterID))
                }
                updateAdapter()
                true
            }
            R.id.chapter_deselect_all -> {
                selectedChapters = ArrayList()
                updateAdapter()
                if (inflater != null) activity?.invalidateOptionsMenu()
                true
            }
            R.id.chapter_mark_read -> {
                for (chapterID in selectedChapters) {
                    if (getChapterStatus(chapterID).a != 2) setChapterStatus(chapterID, Status.READ)
                }
                updateAdapter()
                true
            }
            R.id.chapter_mark_unread -> {
                for (chapterID in selectedChapters) {
                    if (getChapterStatus(chapterID).a != 0) setChapterStatus(chapterID, Status.UNREAD)
                }
                updateAdapter()
                true
            }
            R.id.chapter_mark_reading -> {
                for (chapterID in selectedChapters) {
                    if (getChapterStatus(chapterID).a != 0) setChapterStatus(chapterID, Status.READING)
                }
                updateAdapter()
                true
            }
            R.id.chapter_select_between -> {
                val min = findMinPosition()
                val max = findMaxPosition()
                var x = min
                while (x < max) {
                    if (!contains(novelFragment!!.novelChapters[x])) {
                        val id = getChapterIDFromChapterURL(novelFragment!!.novelChapters[x].link)
                        selectedChapters.add(id)
                    }
                    x++
                }
                updateAdapter()
                true
            }
            R.id.chapter_filter -> {
                novelFragment!!.novelChapters = novelFragment!!.novelChapters.reversed()
                reversed = !reversed
                return updateAdapter()
            }
            else -> false
        }
    }

    operator fun contains(novelChapter: Novel.Chapter): Boolean {
        for (n in selectedChapters) if (getChapter(n)!!.link.equals(novelChapter.link, ignoreCase = true)) return true
        return false
    }

    private fun findMinPosition(): Int {
        var min: Int = novelFragment!!.novelChapters.size
        for (x in novelFragment!!.novelChapters.indices) if (contains(novelFragment!!.novelChapters[x])) if (x < min) min = x
        return min
    }

    private fun findMaxPosition(): Int {
        var max = -1
        for (x in novelFragment!!.novelChapters.indices.reversed()) if (contains(novelFragment!!.novelChapters[x])) if (x > max) max = x
        return max
    }

    private lateinit var receiver: BroadcastReceiver

}